File: bench/agent-explore/driver.mjs
#!/usr/bin/env node
import { readFile, readdir, realpath, stat } from "node:fs/promises";
import { relative, resolve, sep } from "node:path";

import {
  Agent,
  tool,
} from "../../sdk/typescript/packages/muzen-sdk/dist/agent.js";

function parseArgs(argv) {
  const options = { readFiles: 5 };
  for (let index = 0; index < argv.length; index += 1) {
    const flag = argv[index];
    const value = argv[index + 1];
    if (flag === "--transport") (options.transport = value), (index += 1);
    else if (flag === "--root") (options.root = value), (index += 1);
    else if (flag === "--base-url") (options.baseUrl = value), (index += 1);
    else if (flag === "--model-base-url")
      (options.modelBaseUrl = value), (index += 1);
    else if (flag === "--read-files")
      (options.readFiles = Number(value)), (index += 1);
    else throw new Error(`unknown argument: ${flag}`);
  }
  if (!["local_runner", "http"].includes(options.transport))
    throw new Error("--transport must be local_runner or http");
  if (!options.root || !options.modelBaseUrl)
    throw new Error("--root and --model-base-url are required");
  if (options.transport === "http" && !options.baseUrl)
    throw new Error("--base-url is required for http transport");
  if (!Number.isInteger(options.readFiles) || options.readFiles < 1)
    throw new Error("--read-files must be positive");
  return options;
}

function inside(root, candidate) {
  return candidate === root || candidate.startsWith(`${root}${sep}`);
}

async function jailed(root, requested) {
  const candidate = await realpath(resolve(root, requested));
  if (!inside(root, candidate))
    throw new Error(`path escapes --root: ${requested}`);
  return candidate;
}

async function filesUnder(root, requested) {
  const base = await jailed(root, requested);
  const found = [];
  async function visit(directory) {
    const entries = await readdir(directory, { withFileTypes: true });
    entries.sort((left, right) => left.name.localeCompare(right.name));
    for (const entry of entries) {
      const lexical = resolve(directory, entry.name);
      if (entry.isDirectory()) await visit(lexical);
      else if (entry.isFile()) {
        const candidate = await realpath(lexical);
        if (inside(root, candidate))
          found.push({
            path: relative(root, candidate).split(sep).join("/"),
            bytes: (await stat(candidate)).size,
          });
      }
    }
  }
  await visit(base);
  return found.sort((left, right) => left.path.localeCompare(right.path));
}

const options = parseArgs(process.argv.slice(2));
const root = await realpath(options.root);
const counts = { fs_list: 0, fs_read: 0, fs_grep: 0 };
const objectInput = (properties, required) => ({
  type: "object",
  properties,
  required,
  additionalProperties: false,
});

const fsList = tool({
  name: "fs_list",
  description: "Recursively list regular files below a repository path.",
  input: objectInput({ path: { type: "string" } }, ["path"]),
  execute: async ({ path }) => {
    counts.fs_list += 1;
    const files = await filesUnder(root, path);
    return { path, files, totalFiles: files.length };
  },
});
const fsRead = tool({
  name: "fs_read",
  description: "Read one UTF-8 repository file.",
  input: objectInput({ path: { type: "string" } }, ["path"]),
  execute: async ({ path }) => {
    counts.fs_read += 1;
    const data = await readFile(await jailed(root, path));
    return { path, bytes: data.length, content: data.toString("utf8") };
  },
});
const fsGrep = tool({
  name: "fs_grep",
  description: "Search repository files for a fixed text pattern.",
  input: objectInput(
    { pattern: { type: "string" }, path: { type: "string" } },
    ["pattern", "path"],
  ),
  execute: async ({ pattern, path }) => {
    counts.fs_grep += 1;
    const matches = [];
    let totalMatches = 0;
    for (const entry of await filesUnder(root, path)) {
      const lines = (
        await readFile(await jailed(root, entry.path), "utf8")
      ).split(/\r?\n/);
      lines.forEach((line, index) => {
        if (line.includes(pattern)) {
          totalMatches += 1;
          if (matches.length < 100)
            matches.push({
              path: entry.path,
              line: index + 1,
              text: line.slice(0, 500),
            });
        }
      });
    }
    return {
      pattern,
      path,
      matches,
      totalMatches,
      truncated: totalMatches > matches.length,
    };
  },
});

const agentOptions = {
  instructions:
    "Explore the repository with the provided filesystem tools, then summarize what you saw.",
  model: "openai:muzen-agent-explore",
  tools: [fsList, fsRead, fsGrep],
  transport: options.transport,
  apiKey: "bench-test-key",
  ...(options.transport === "local_runner"
    ? { baseUrl: options.modelBaseUrl }
    : { baseUrl: options.baseUrl, modelBaseUrl: options.modelBaseUrl }),
};

const started = performance.now();
const agent = new Agent(agentOptions);
let result;
try {
  result = (
    await agent.run("Explore src and report a concise summary.")
  ).raiseForStatus();
} finally {
  await agent.close();
}
const expected = { fs_list: 1, fs_read: options.readFiles, fs_grep: 1 };
if (JSON.stringify(counts) !== JSON.stringify(expected))
  throw new Error(
    `tool count mismatch: expected ${JSON.stringify(expected)}, got ${JSON.stringify(counts)}`,
  );
if (!result.text.trim()) throw new Error("agent returned an empty summary");
console.log(
  JSON.stringify({
    turns: Object.values(counts).reduce((sum, value) => sum + value, 1),
    toolCalls: Object.values(counts).reduce((sum, value) => sum + value, 0),
    durationMs: Number((performance.now() - started).toFixed(3)),
    summaryText: result.text,
  }),
);

File: bench/agent-explore/driver.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import asyncio
import json
import os
import time
from pathlib import Path
from threading import Lock
from typing import Any

from muzen import Agent, tool


def jailed(root: Path, requested: str) -> Path:
    candidate = (root / requested).resolve()
    if candidate != root and root not in candidate.parents:
        raise ValueError("path escapes --root: %s" % requested)
    return candidate


def files_under(root: Path, requested: str) -> list[dict[str, Any]]:
    base = jailed(root, requested)
    entries = []
    for path in sorted(item for item in base.rglob("*") if item.is_file()):
        resolved = path.resolve()
        if resolved != root and root not in resolved.parents:
            continue
        entries.append({"path": resolved.relative_to(root).as_posix(), "bytes": resolved.stat().st_size})
    return entries


async def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--transport", choices=("local_runner", "http"), required=True)
    parser.add_argument("--root", type=Path, required=True)
    parser.add_argument("--base-url")
    parser.add_argument("--model-base-url", required=True)
    parser.add_argument("--read-files", type=int, default=5)
    args = parser.parse_args()
    root = args.root.resolve()
    if args.transport == "http" and not args.base_url:
        parser.error("--base-url is required for http transport")
    if args.read_files < 1:
        parser.error("--read-files must be positive")

    counts = {"fs_list": 0, "fs_read": 0, "fs_grep": 0}
    count_lock = Lock()

    @tool
    def fs_list(path: str) -> dict[str, Any]:
        """Recursively list regular files below a repository path."""
        with count_lock:
            counts["fs_list"] += 1
        entries = files_under(root, path)
        return {"path": path, "files": entries, "totalFiles": len(entries)}

    @tool
    def fs_read(path: str) -> dict[str, Any]:
        """Read one UTF-8 repository file."""
        with count_lock:
            counts["fs_read"] += 1
        target = jailed(root, path)
        data = target.read_bytes()
        return {"path": path, "bytes": len(data), "content": data.decode("utf-8", errors="replace")}

    @tool
    def fs_grep(pattern: str, path: str) -> dict[str, Any]:
        """Search repository files for a fixed text pattern."""
        with count_lock:
            counts["fs_grep"] += 1
        matches: list[dict[str, Any]] = []
        total = 0
        for entry in files_under(root, path):
            target = jailed(root, entry["path"])
            for number, line in enumerate(target.read_text(encoding="utf-8", errors="replace").splitlines(), 1):
                if pattern in line:
                    total += 1
                    if len(matches) < 100:
                        matches.append({"path": entry["path"], "line": number, "text": line[:500]})
        return {"pattern": pattern, "path": path, "matches": matches, "totalMatches": total, "truncated": total > len(matches)}

    options: dict[str, Any] = {
        "instructions": "Explore the repository with the provided filesystem tools, then summarize what you saw.",
        "model": "openai:muzen-agent-explore",
        "tools": [fs_list, fs_read, fs_grep],
        "transport": args.transport,
        "api_key": "bench-test-key",
        "model_base_url": args.model_base_url,
    }
    if args.transport == "local_runner":
        options["base_url"] = args.model_base_url
        options.pop("model_base_url")
    else:
        options["base_url"] = args.base_url

    started = time.perf_counter()
    agent = Agent(**options)
    try:
        result = await agent.run("Explore src and report a concise summary.")
        result.raise_for_status()
    finally:
        await agent.close()
    duration_ms = round((time.perf_counter() - started) * 1000, 3)
    expected = {"fs_list": 1, "fs_read": args.read_files, "fs_grep": 1}
    if counts != expected:
        raise RuntimeError("tool count mismatch: expected %r, got %r" % (expected, counts))
    if not result.text.strip():
        raise RuntimeError("agent returned an empty summary")
    print(json.dumps({
        "turns": sum(counts.values()) + 1,
        "toolCalls": sum(counts.values()),
        "durationMs": duration_ms,
        "summaryText": result.text,
    }, separators=(",", ":")))


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except Exception as error:
        print("muzen agent explore Python driver: %s" % error, file=os.sys.stderr)
        raise SystemExit(1)
